programs/pkey/key_app_writer.c

```diff
@@ -192,15 +192,26 @@ static int write_private_key( mbedtls_pk_context *key, const char *output_file )
 
 int main( int argc, char *argv[] )
 {
-    int ret = 1;
+    int ret = 0;
     int exit_code = MBEDTLS_EXIT_FAILURE;
     char buf[1024];
     int i;
     char *p, *q;
 
+#if defined(MBEDTLS_PLATFORM_C)
+    mbedtls_platform_context platform_ctx;
+#endif
     mbedtls_pk_context key;
     mbedtls_mpi N, P, Q, D, E, DP, DQ, QP;
 
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n", -ret );
+        return( 1 );
```

Comment:
This place is an example of inconsistent return/exit handling. Please compare this to line 424:
`if( exit_code != MBEDTLS_EXIT_SUCCESS && ret != 0 )`
I can't see a good reason for having 3 places to handle returning an error. 

What I suggest is to get rid of the "exit_code" variable in this whole PR, leave the "ret" variable, and, if we're not overwriting it with MBEDTLS_EXIT_SUCCESS or MBEDTLS_EXIT_FAILURE by hand entirely in the scope of a program (thus passing it outside) - add a bound check in the `exit:` stage.